Add ability to compute residual and jacobian together

lindsayad opened the pull request:
closes #19444 

To-do:
- [x] make work with at least CFEM (currently works with FV)
- [x] make work with PJFNK (pending libmesh/libmesh#3162)
- [x] make work with auto scaling
- [x] Extend `NonlinearThread` refactor to `ComputeResidualAndJacobian`

lindsayad committed:
Add ability to compute residual and Jacobian together

Closes #19444

lindsayad committed:
Add new API in Assembly that handles both residual and Jacobian

This allows user code to be simpler

lindsayad committed:
Update NSFV code to be able to compute resid and Jac together

For the mixing length wcnsfv channel flow test, after refining the mesh
twice, computing together reduces run time from 75 to 66 seconds, a 12%
reduction

lindsayad committed:
Implement computeResidualAndJacobian for ADKernel

lindsayad committed:
Implement computeResidualAndJacobian for ADNodalBC

lindsayad committed:
Enable resid-jac-together for ADIntegratedBC

And also make sure that we are updating our variable and material
property dependencies. This was necessary in order to make a tensor
mechanics plasticity test run successfully

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/956050/) on 43767ee wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19971/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19971/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 23a1b44196b378faee28ac24f3409520a6509768`

lindsayad committed:
clang-format

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/956137/) on c709764 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/19971/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/956147/) on c709764 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>bbd401</th>
      <th colspan=4>#19971 c70976</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.32%</td>
      <td>82.29%</td>
      <td>-0.03%</td>
      <td>80.36%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>73668</td>
      <td>74123</td>
      <td>+455</td>
      <td>532</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15822</td>
      <td>15954</td>
      <td>+132</td>
      <td>130</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19971/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19971/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>bbd401</th>
      <th colspan=4>#19971 c70976</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.55%</td>
      <td>83.51%</td>
      <td>-0.03%</td>
      <td>57.14%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>8257</td>
      <td>8252</td>
      <td>-5</td>
      <td>4</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1626</td>
      <td>1629</td>
      <td>+3</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19971/coverage_diff/navier_stokes.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19971/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/19971/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 80.36% is less than the suggested 90.0%
- `navier_stokes` new line coverage rate 57.14% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad committed:
FVFluxBC::computeResidualAndJacobian testing

lindsayad committed:
implement/test FVScalarLagrangeMultiplierInterface resid and Jac

lindsayad committed:
FVInterfaceKernel::computeResidualAndJacobian testing

lindsayad committed:
Implement and test mortar resid and Jacobian together

Also make sure to test on a displaced mesh configuration

loganharbour reviewed (commented):
this seems like an insane amount of duplication... is it really required?

lindsayad commented:
insane seems a bit superlative

loganharbour commented:
maybe egregious? still too much?

lindsayad committed:
Mirror processDerivatives in ADNodalKernel

lindsayad commented:
I would say that I worked decently hard to reduce a lot of potential duplication, so I don't really appreciate the language. I will try to consolidate more and modify some `computeJacobians` such that `compouteResidualAndJacobian` can just call those

lindsayad commented on framework/src/bcs/ADIntegratedBC.C:
```diff
@@ -139,6 +139,14 @@ ADIntegratedBCTempl<T>::computeResidualsForJacobian()
 
+template <typename T>
+void
+ADIntegratedBCTempl<T>::computeResidualAndJacobian()
+{
+  computeResidualsForJacobian();
+  _assembly.processResiduals(_residuals, _var.dofIndices(), _vector_tags, _matrix_tags);
+}
```
aka here

loganharbour commented on framework/src/loops/ComputeResidualAndJacobianThread.C:
```diff
@@ -0,0 +1,173 @@
+
+#include "libmesh/threads.h"
+
+ComputeResidualAndJacobianThread::ComputeResidualAndJacobianThread(
```
sorry - not trying to be rude.

this for example looks like it duplicates quite a lot from the standard residual/jacobian loop, and also misses quite a bit of functionality. with such duplication, it becomes easier and easier to forget about objects when adding features

lindsayad replied:
Some of the object forgetting can be purposeful 😄 But it is definitely possible to forget by mistake although I've tried to implement a good error message that would make it easy to know what is missing

There is a lot of duplication between this and `ComputeJacobianThread` and `ComputeResidualThread`. You can pretty easily remove some duplication in things like `subdomainChanged`. But it's harder for the `onElement`, `onBoundary`, `onInternalSide`, etc. methods because at the core they have to call `computeResidual` or `computeJacobian` or `computeResidualAndJacobian`. I don't think we want to go down the road of changing those user interfaces to something like `compute(Moose::ComputeType)`. Maybe the thing to do would be to do what I did in `ComputeFVFluxThread` which is create a generic loop with a `virtual compute(ResidualObject &)` that derived classes like `ComputeResidualThread` override to dispatch to `computeResidual()` via like
```c++
ComputeResidualThread::compute(ResidualObject & ro)
{
  ro.computeResidual();
}
```

lindsayad replied:
☹️ even for kernels, there are subtle differences
`ComputeResidualThread`:
```c++
  if (_tag_kernels->hasActiveBlockObjects(_subdomain, _tid))
  {
    const auto & kernels = _tag_kernels->getActiveBlockObjects(_subdomain, _tid);
    for (const auto & kernel : kernels)
      kernel->computeResidual();
  }
```
`ComputeJacobianThread`:
```c++
  if (_warehouse->hasActiveBlockObjects(_subdomain, _tid))
  {
    auto & kernels = _warehouse->getActiveBlockObjects(_subdomain, _tid);
    for (const auto & kernel : kernels)
      if (kernel->isImplicit())
      {
        kernel->prepareShapes(kernel->variable().number());
        kernel->computeJacobian();
	/// done only when nonlocal kernels exist in the system                                                                                                                                                                               
        if (_fe_problem.checkNonlocalCouplingRequirement())
        {
          std::shared_ptr<NonlocalKernel> nonlocal_kernel =
              std::dynamic_pointer_cast<NonlocalKernel>(kernel);
          if (nonlocal_kernel)
            kernel->computeNonlocalJacobian();
        }
      }
  }
```

loganharbour commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -5329,6 +5329,164 @@ FEProblemBase::computeResidual(const NumericVector<Number> & soln, NumericVector
 }
 
+void
+FEProblemBase::computeResidualAndJacobian(const NumericVector<Number> & soln,
```
same thing here; I fear a maintenance burden. but at the same time... may be impossible to clean up something like this as the residual/jacobian method madness in FEProblemBase is insane 😆

lindsayad replied:
Yea this one is even tougher in my mind ... I honestly feel like there may be some inefficiencies in the current methods. Or more importantly, there are enough differences between the `computeResidual` and `computeJacobian` paths that without starting from scratch, working with objects for whom I know how they work, I fear I would miss some nuanced difference in how the residual and Jacobian have to be treated that could cause some difficult bugs to track down. Thus far I have mostly stuck with AD objects for whom I understand clearly how they work

lindsayad committed:
Code consolidation

lindsayad committed:
Move residual and jac initialization back to constructor

I don't know why I did this in the first place. If it turns out there
wasn't a good reason, then this can be squashed

lindsayad committed:
Allow computeResidualAndJacobian to just call computeJacobian

lindsayad committed:
Scaling testing with computing resid and Jac together

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 544e2c87ea2f35aab900dde8b8f1e0926d720421
+Subproject commit ffc792cf61a21d470bb181bead03c6b5100cbe7a
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Controlled App tests](https://civet.inl.gov/job/958380/) on 50e8b20 : invalidated by @lindsayad

lindsayad committed:
Really do compute the residual and Jacobian together

It turns out that the movement of residual and Jacobian object setting
from the constructor to init was very intentional. This is because
FEProblemSolve construction has not occurred at the time of nonlinear
system construction. With this new commit, when FEProblemSolve calls
NonlinearSystemBase::residAndJacobianTogether we toggle from using
tradtional residual and Jacobian to the joint object. This should be
compatible with PBP which were the tests that failed when I had moved
object setting to init.

This commit also introduces a new processResiduals API meant only for
processing residuals (and not their derivatives/Jacobian). I did this
because I was getting failures in NDEBUG modes about different residual
evaluations between the JFNK residual evaluation and the nonlinear residual
evaluation. This was due to the fact that we were caching per element
per residual object for the nonlinear residual and per element (all
residual objects) for JFNK residual. The new processResiduals API allows
residual objects to follow similar patterns for their residual and Jacobian
evaluation implementations.

When I added the new processResiduals, I didn't initially restrict it to
global AD indexing and I got an error on CIVET concerning the lack of
definition of `_scaling_vector` in the processResiduals implementation.
This led to me adding a `scaling_factor` argument to call the vector
of residuals APIs I have on Assembly, such that the methods can be used
equally well by both non-global and global AD. I'm not ready to give up
totally on non-global, nonsparse AD because Gavin has shown it's
demonstrably faster than global, sparse for our INSAD navier-stokes
implementation.

This regolds a THM test that was previously failing at time step 334. I
don't think that my changes in this commit made the solve any better; I
think this test was just sitting right on the edge

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/961786/) on 6dd6390 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/961786/) on 6dd6390 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/962025/) on 7478e9f wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/962025/) on 7478e9f wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

lindsayad commented:
Ok I don't think I'm going to attempt to consolidate the `FEProblemBase/NonlinearSystemBase` method code, but I do think I can factor out a lot of stuff in the loop code into a class which I think I'll call `NonlinearThread`. @loganharbour would this appease you 😄 ?

loganharbour commented:
> Ok I don't think I'm going to attempt to consolidate the `FEProblemBase/NonlinearSystemBase` method code, but I do think I can factor out a lot of stuff in the loop code into a class which I think I'll call `NonlinearThread`. @loganharbour would this appease you 😄 ?

I can get behind new features + some cleanup any day of the week

lindsayad committed:
Create NonlinearThread

lindsayad commented:
Don’t close

> On Mar 24, 2022, at 10:43 PM, github-actions[bot] ***@***.***> wrote:
> 
> ﻿
> This pull request has been automatically marked as stale because it has not had recent activity in the last 30 days. It will be closed in 3 days if no further activity occurs. Thank you for your contributions.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

lindsayad committed:
Refactor FVKernel to inherit from ResidualObject

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/999052/) on 2b7c069 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19971/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19971/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c90b1f04c38b5fa8c35d8b16e7ca4ae7c7ad7115`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/999052/) on 2b7c069 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/999052/) on 2b7c069 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/999083/) on 9b373c7 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19971/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19971/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c90b1f04c38b5fa8c35d8b16e7ca4ae7c7ad7115`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/999083/) on 9b373c7 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/999083/) on 9b373c7 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/999114/) on 81540a5 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/999114/) on 81540a5 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/999633/) on 42f7293 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/999633/) on 42f7293 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/999881/) on e25c900 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1000036/) on c940e26 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1000036/) on c940e26 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

lindsayad commented:
I'm deciding not to pursue refactoring `ComputeJacobianThread` to inherit from `NonlinearThread`, at least not at this time. There's quite a bit of stuff unique to that code

loganharbour commented:
> I'm deciding not to pursue refactoring `ComputeJacobianThread` to inherit from `NonlinearThread`, at least not at this time. There's quite a bit of stuff unique to that code

Sounds fine to me.

moosebuild commented:
Job [Controlled app tests](https://civet.inl.gov/job/1006039/) on 33694c9 : invalidated by @lindsayad

lindsayad commented:
I think this is ready for review

lindsayad committed:
Test lower-d var coupling into bcs

lindsayad committed:
Generalize LowerDBlockFromSidesetGenerator to work with distributed mesh

lindsayad committed:
Loosen absolute tolerance for imperfect gas test

The change in FP operations for AD kernel residual caching seems to
have pushed this on-the-edge test over the edge. Loosening the
solver absolute tolerance seems to allow the test to still yield
the same results

loganharbour commented on framework/include/base/Assembly.h:
```diff
@@ -1688,6 +1688,15 @@ class Assembly
                           dof_id_type dof_index,
                           const std::set<TagID> & matrix_tags);
 
+  /**
+   * This method is only meant to be called if MOOSE is configured to use global AD indexing. This
+   * performs the duties of both \p processResidual and \p processDerivatives
+   */
```
Should be able to be ifdef-ed?

loganharbour commented on framework/include/loops/ComputeFVFluxThread.h:
```diff
@@ -362,36 +379,23 @@ ComputeFVFluxThread<RangeType>::reinitVariables(const FaceInfo & fi)
-      k->computeJacobian(fi);
-    else
-      k->computeResidual(fi);
+  for (auto * const k : _fv_flux_kernels)
```
not your doing, but IMO let's nuke the early return above. your call

loganharbour commented on framework/src/base/ResidualObject.C:
```diff
@@ -63,3 +63,14 @@ ResidualObject::prepareShapes(const unsigned int var_num)
+void
+ResidualObject::computeResidualAndJacobian()
+{
+  mooseError("'",
+             name(),
+             "' is of system type '",
```
`MooseObject::mooseError()` has the `type()` and `name()` context

loganharbour reviewed (changes requested):
All minor changes. Looks good. The loop cleanup is also 👍🏼 

Do we have any corner cases for unimplemented `computeResidualAndJacobian()`?

loganharbour commented on framework/src/loops/ComputeResidualAndJacobianThread.C:
```diff
@@ -0,0 +1,117 @@
+}
+
+void
+ComputeResidualAndJacobianThread::accumulateNeighbor()
```
This is the only thing missing coverage that stuck out to me

lindsayad commented on framework/include/base/Assembly.h:
```diff
@@ -2800,63 +2861,52 @@ void
 
-    const auto & sparse_derivatives = residuals[i].derivatives();
+  // Need to make a copy because we might modify this in constrain_element_vector
+  std::vector<dof_id_type> row_indices = input_row_indices;
```
Because we might modify the vector in `constrain_element_vector` and the input indices correspond to the unconstrained dof indices, which is data we don't want to modify. Let me know if I can make the comment clearer

lindsayad committed:
Logan review: ifdefs and renaming

- try to make sure that any code that would generate a run-time error
  instead is just not available for compile
- rename processResidual(s) to processResidual(s)AndDerivatives for
  ADReal overloads

lindsayad commented on python/TestHarness/testers/PetscJacobianTester.py:
```diff
@@ -18,7 +18,7 @@ class PetscJacobianTester(RunApp):
         params = RunApp.validParams()
         params.addParam('ratio_tol', 1e-8, "Relative tolerance to compare the ration against.")
-        params.addParam('difference_tol', 1e-8, "Relative tolerance to compare the difference against.")
+        params.addParam('difference_tol', 1e-7, "Relative tolerance to compare the difference against.")
```
Users of `PetscJacobianTester` basically always have to manually set a value for `difference_tol` currently because 1e-8 is too tight. 1e-7 is probably too tight as well, but at least it's a little better. In practice, it is the `ratio_tol` that really matters for communicating whether the hand-coded (or AD) Jacobian is accurate. We should probably consider removing this parameter altogether

lindsayad committed:
process.*Derivatives -> process.*Jacobian

lindsayad committed:
Create _residuals/_residuals_and_jacobians members in ADIntegratedBC

lindsayad committed:
documentation and newsletter entry

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1069878/) on 9b6a184 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19971/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19971/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d444c549368b641223dca5ffd90e07d934b1f99a`

lindsayad commented:
Ok @loganharbour I think this is ready for further review

loganharbour reviewed (dismissed):
Gave it another run through; resolved comments that were resolved by your last few commits. 🍻 

Word of note: some people may be frustrated with the use of `resid` instead of `residual` in `resid_and_jacobian_together`, but I am not one of those people.

I'm going to add some more testing to this.

loganharbour commented:
Aha... looks like adding those extra tests wasn't the worst idea here 😆

lindsayad committed:
resid_and_jacobian -> residual_and_jacobian

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1070667/) on eb50bac wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19971/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19971/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 65228e22d11d97b0b355178c64ae457c350f8134`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1070724/) on eac25ab wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19971/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19971/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 65228e22d11d97b0b355178c64ae457c350f8134`

lindsayad committed:
distributed consistency with just ghosted boundary elems

lindsayad committed:
Slightly bump ratio tol for scaling Jacobian tests

lindsayad commented:
@loganharbour look good?

recuero commented on framework/include/utils/MortarUtils.h:
```diff
@@ -97,14 +97,23 @@ loopOverMortarSegments(const Iterators & secondary_elems_to_mortar_segments,
   // The element Jacobian times weights
   const auto & JxW_msm = assembly.jxWMortar();
 
+  // Set required variables
+  std::set<MooseVariableFieldBase *> needed_moose_vars;
+  for (const auto & consumer : consumers)
+  {
+    const auto & mv_deps = consumer->getMooseVariableDependencies();
+    needed_moose_vars.insert(mv_deps.begin(), mv_deps.end());
+  }
+  fe_problem.setActiveElementalMooseVariables(needed_moose_vars, /*tid=*/tid);
+
```
@lindsayad it appears these additional `fe_problem` dependencies are causing early failures of some assessment cases that use mortar: Any ideas why this could be?
